tools/app.hpp

```diff
@@ -124,11 +124,33 @@ namespace arg {
         std::string m_algorithm;
     };
 
+    enum class ScorerMode : bool { Required, Optional };
+
+    template <typename T>
+    void add_scorer_options(CLI::App* app, T& args, bool scorer_required)
+    {
+        if (scorer_required) {
+            app->add_option("-s,--scorer", args.m_params.name, "Scorer function")->required();
+        } else {
+            app->add_option("-s,--scorer", args.m_params.name, "Scorer function");
+        }
+
+        app->add_option("--bm25-k1", args.m_params.bm25_k1, "BM25 k1 parameter.");
+        app->add_option("--bm25-b", args.m_params.bm25_b, "BM25 b parameter.");
+        app->add_option("--pl2-c", args.m_params.pl2_c, "PL2 c parameter.");
+        app->add_option("--qld-mu", args.m_params.qld_mu, "QLD mu parameter.");
+    }
+
+    template <ScorerMode Mode = ScorerMode::Required>
     struct Quantize {
         explicit Quantize(CLI::App* app) : m_params("")
         {
             auto* wand = app->add_option("-w,--wand", m_wand_data_path, "WAND data filename");
-            add_scorer_options(app, *this);
+            if constexpr (Mode == ScorerMode::Required) {
+                add_scorer_options(app, *this, true);
+            } else {
+                add_scorer_options(app, *this, false);
+            }
```

Comment:
Let's make `add_scorer_options` take `ScorerMode` instead of `bool` and just pass it directly here `add_scorer_options(app, *this, Mode);`.

Then, in the other function `if (scorer_mode == ScorerMode::Required)` without `constexpr`.